### DISCUSSION TITLE
Globally subscribe to actions

### DISCUSSION THREAD

**Discussion by phil294** [vuejs/pinia#1037] [Help and Questions] [answered]
Trying to subscribe to any action call from *any* store. Use case: Logging for automated error reports (store action history: very useful information).

While Vuex had a global `store.subscribe()` hook, in Pinia, afaik the only solution is using a plugin:
```javascript
pinia.use({ store } => {
    store.$onAction({ name, args } => /* do things */)
```

And this works, but when in combination with Vue Router and Options API, this yields in `onUnmounted is called when there is no active component instance to be associated with. Lifecycle injection APIs can only be used during execution of setup(). If you are using async setup(), make sure to register lifecycle hooks before the first await statement.` upon any route navigation. How do you do this?

Edit: If this is unexpected / a bug, please tell me so I can fiddle together a repro repo, but I'd prefer not to as it is a complex setup that needs a lot of narrowing down

**Comment by posva** [comment, marked as answer]
Using a plugin is the correct way due to having multiple stores in Pinia.

the `onMounted()` warning cannot come from pinia as it doesn't have any occurrence in its code

**Reply by phil294** [reply]
> Using a plugin is the correct way due to having multiple stores in Pinia.
> 
> the `onMounted()` warning cannot come from pinia as it doesn't have any occurrence in its code

ok
yeah, it's a Vue3 message that is triggered by the `$onAction` call.

**Reply by posva** [reply]
Ah, I misread. `onUnmounted()` is used by subscriptions (https://github.com/vuejs/pinia/blob/fe356ec5e0807b8921d3a1cfd0155ac094feb9f4/packages/pinia/src/subscriptions.ts#L22-L24). Are you using `await` in the relevant pinia plugin? If not, a boiled down reproduction would indeed help!
